packages/react/src/hooks/program.ts

```diff
@@ -0,0 +1,200 @@
+// TODO: Replace any with inferred types.
+
+"use client";
+
+import {
+  useMutation,
+  type UseMutationOptions,
+  useQuery,
+  useQueryClient,
+  type UseQueryOptions,
+} from "@tanstack/react-query";
+import type { Address, Instruction } from "gill";
+import {
+  createSolanaClient,
+  createTransaction,
+  type Signature,
+  type SolanaClient,
+  type TransactionSendingSigner,
+} from "gill";
+
+import { GILL_HOOK_CLIENT_KEY } from "../const.js";
+
+type InstructionBuilder<TInput = any> = (input: TInput, config?: { programAddress: Address }) => Instruction;
+
+type AccountFetcher<TData = any> = (rpc: SolanaClient["rpc"], address: Address, config?: any) => Promise<TData>;
+
+type CommitmentLevel = "confirmed" | "finalized";
+
+type SignAndSendFn = (
+  instruction: Instruction,
+  signer: TransactionSendingSigner,
+) => Promise<string>;
+
+type ProgramHookConfig<
+  TInstructions extends Record<string, InstructionBuilder<any>>,
+  TAccounts extends Record<string, AccountFetcher<any>>,
+> = {
+  accounts: TAccounts;
+  commitment?: CommitmentLevel | null;
+  instructions: TInstructions;
+  programAddress: Address;
+};
+
+type UseProgramMutationInput<
+  TInstructions extends Record<string, InstructionBuilder<any>>,
+  TInstructionName extends keyof TInstructions,
+> = Omit<
+  UseMutationOptions<
+    string,
+    Error,
+    {
+      commitment?: CommitmentLevel | null;
+      params: Parameters<TInstructions[TInstructionName]>[0];
+      rpc: SolanaClient['rpc'];
+      signAndSend: SignAndSendFn;
+      signer: TransactionSendingSigner;
+    }
+  >,
+  "mutationFn"
+> & {
+  commitment?: CommitmentLevel | null;
+  instruction: TInstructionName;
+};
+
+type UseProgramQueryInput<
+  TAccounts extends Record<string, AccountFetcher<any>>,
+  TAccountName extends keyof TAccounts,
+> = Omit<UseQueryOptions<Awaited<ReturnType<TAccounts[TAccountName]>>, Error>, "queryFn" | "queryKey"> & {
+  account: TAccountName;
+  address: Address;
+  rpc: Parameters<TAccounts[TAccountName]>[0];
+};
+
+const CONFIRM_TRANSATION_MAX_RETRIES = 15;
+const CONFIRM_TRANSATION_COMMITMENT_FINALIZED_MILLISECOND_DELAY = 2000;
+const CONFIRM_TRANSATION_COMMITMENT_CONFIRMED_MILLISECOND_DELAY = 100;
+const CONFIRM_TRANSATION_MAX_MILLISECOND_DELAY = 2000;
```

Comment:
These constants for confirming transaction were found through trial and error. We can adjust these if needed. I gave confirmed an initial shorter delay with exponential backoff but for finalized transactions, it just requests every 2 seconds.